pkg/network/encoding/usm.go

```diff
@@ -0,0 +1,186 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package encoding
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/network"
+	"github.com/DataDog/datadog-agent/pkg/network/protocols/telemetry"
+	"github.com/DataDog/datadog-agent/pkg/network/types"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// USMConnectionIndex provides a generic container for USM data pre-aggregated by connection
+type USMConnectionIndex[K comparable, V any] struct {
+	lookupFn func(network.ConnectionStats, map[types.ConnectionKey]*USMConnectionData[K, V]) *USMConnectionData[K, V]
+	data     map[types.ConnectionKey]*USMConnectionData[K, V]
+	protocol string
+	once     sync.Once
+}
+
+// USMConnectionData aggregates all USM data associated to a specific connection
+type USMConnectionData[K comparable, V any] struct {
+	Data []USMKeyValue[K, V]
+
+	// This is used for handling PID collisions
+	// See notes in `IsPIDCollision`
+	sport, dport uint16
+
+	// Used for the purposes of orphan aggregation count
+	claimed bool
+
+	// Used during the first pass to determine the size of the `Data`
+	size int
+}
+
+type USMKeyValue[K comparable, V any] struct {
+	Key   K
+	Value V
+}
+
+// GroupByConnection generates a `USMConnectionIndex` from a generic `map[K]V` data structure.
+// In addition to the `data` argument the caller must provide a `keyGen` function that
+// essentially translates `K` to a `types.ConnectionKey` and a `protocol` name.
+func GroupByConnection[K comparable, V any](protocol string, data map[K]V, keyGen func(K) types.ConnectionKey) *USMConnectionIndex[K, V] {
+	byConnection := &USMConnectionIndex[K, V]{
+		lookupFn: USMLookup[K, V],
+		data:     make(map[types.ConnectionKey]*USMConnectionData[K, V], len(data)/2),
```

Comment:
I haven't thought too hard about  it to be honest, since it's a bit hard to predict the size of this map, but obviously the intent here was to amortize the cost of re-sizing the map too many times.

The original map `data` is keyed more or less like `(connection tuple, HTTP path, HTTP method)` in the particular context of HTTP (though as reminder this is a generic data container, so we can't rely too much on the specifics about HTTP).

The generated map is keyed by `connection tuple`, so the cardinality of the generated map is less or equal than the original map cardinality.

The difference between the two is dependent on the workload. For example, if you have one HTTP client issuing requests to 100 different endpoints to the same server from the same TCP connection, then cardinality of the original map would be `100` and of the generated map would be `1`.

I would expect _some_ level of aggregation in most workloads, so I chose `len(data)/2`.

An alternative approach would be iterating over all `network.ConnectionStats` and counting how many of them are classified as HTTP. But that has its own problems as well (classification might be disabled,  there may be a mismatch between what NPM and USM sees etc)

Reply:
let's write it in the code, to ensure we're preserving the knowledge about that

Reply:
Done 👍